src/system/SystemLayerImplZephyr.cpp

```diff
@@ -0,0 +1,235 @@
+
+    CHIP_SYSTEM_FAULT_INJECT(FaultInjection::kFault_TimeoutImmediate, delay = System::Clock::kZero);
+
+    CancelTimer(onComplete, appState);
+
+    TimerList::Node * timer = mTimerPool.Create(*this, SystemClock().GetMonotonicTimestamp() + delay, onComplete, appState);
+    VerifyOrReturnError(timer != nullptr, CHIP_ERROR_NO_MEMORY);
+
+    if (mTimerList.Add(timer) == timer)
+    {
+        // The new timer is the earliest, so the time until the next event has probably changed.
+        TriggerPlatformTimerUpdate();
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR LayerImplZephyr::ExtendTimerTo(Clock::Timeout delay, TimerCompleteCallback onComplete, void * appState)
+{
+    VerifyOrReturnError(delay.count() > 0, CHIP_ERROR_INVALID_ARGUMENT);
+
+    assertChipStackLockedByCurrentThread();
+
+    Clock::Timeout remainingTime = mTimerList.GetRemainingTime(onComplete, appState);
+    if (remainingTime.count() < delay.count())
+    {
+        if (remainingTime == Clock::kZero)
+        {
+            // If remaining time is Clock::kZero, it might possible that our timer is in
+            // the mExpiredTimers list and about to be fired. Remove it from that list, since we are extending it.
+            mExpiredTimers.Remove(onComplete, appState);
+        }
+        return StartTimer(delay, onComplete, appState);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+bool LayerImplZephyr::IsTimerActive(TimerCompleteCallback onComplete, void * appState)
+{
+    bool timerIsActive = (mTimerList.GetRemainingTime(onComplete, appState) > Clock::kZero);
+
+    if (!timerIsActive)
+    {
+        // check if the timer is in the mExpiredTimers list about to be fired.
+        for (TimerList::Node * timer = mExpiredTimers.Earliest(); timer != nullptr; timer = timer->mNextTimer)
+        {
+            if (timer->GetCallback().GetOnComplete() == onComplete && timer->GetCallback().GetAppState() == appState)
+            {
+                return true;
+            }
+        }
+    }
+
+    return timerIsActive;
+}
+
+Clock::Timeout LayerImplZephyr::GetRemainingTime(TimerCompleteCallback onComplete, void * appState)
+{
+    return mTimerList.GetRemainingTime(onComplete, appState);
+}
+
+void LayerImplZephyr::CancelTimer(TimerCompleteCallback onComplete, void * appState)
+{
+    assertChipStackLockedByCurrentThread();
+
+    VerifyOrReturn(mLayerState.IsInitialized());
+
+    TimerList::Node * timer = mTimerList.Remove(onComplete, appState);
+    if (timer == nullptr)
+    {
+        // The timer was not in our "will fire in the future" list, but it might
+        // be in the "we're about to fire these" chunk we already grabbed from
+        // that list.  Check for it there too, and if found there we still want
+        // to cancel it.
+        timer = mExpiredTimers.Remove(onComplete, appState);
+    }
+
+    VerifyOrReturn(timer != nullptr);
+    mTimerPool.Release(timer);
+
+    TriggerPlatformTimerUpdate();
```

Comment:
This seems unnecessary - even if the CHIP thread waits for N seconds and then the timer is already gone, this is fine and we don't have to wait it sooner to figure this out. Especially that the timer may then be rescheduled and wake the thread again.